### PR TITLE
Convert javascript practice to typescript practice

### PR DESCRIPTION
**Pull request by Canh2810**

### CODE REVIEW

File: typescript-practice/.vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+  "editor.formatOnSave": true,
```
Comment (MinhNguyen41092) [inline typescript-practice/.vscode/settings.json]: Don't commit this folder

File: typescript-practice/package.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  "source": "./src/index.html",
+  "scripts": {
+    "start": "parcel",
+    "build": "parcel build"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "devDependencies": {
+    "nodemon": "^2.0.19",
```
Comment (MinhNguyen41092) [inline typescript-practice/package.json]: Why do you need nodemon and ts-node?

File: typescript-practice/src/ts/interface/product-interface.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export interface ProductObj {
```
Comment (MinhNguyen41092) [inline typescript-practice/src/ts/interface/product-interface.ts]: Interface name should be `I<ObjectName>`. For example: `IProduct`

File: typescript-practice/src/ts/controllers/product-controllers.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import { ProductModel } from '../models/product-model';
+import { ProductView } from '../views/product-view';
+/**
+ * @class ProductController
+ *
+ * Links the user input and the view output.
+ *
+ */
+export class ProductController {
```
Comment (MinhNguyen41092) [inline typescript-practice/src/ts/controllers/product-controllers.ts]: All classes should have an interface.

File: typescript-practice/src/ts/views/product-view.ts [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    });
+  }
+
+  // Validate input in form modal
+  validateInput() {
+    if (this.formInput) {
+      this.formInput.forEach(function (inputElement) {
+        if (inputElement) {
+          const formGroup = inputElement.parentElement as HTMLElement;
+          const errorElement = formGroup.querySelector('.form-message') as HTMLElement;
```
Comment (MinhNguyen41092) [inline typescript-practice/src/ts/views/product-view.ts:109]: Why don't you get error message element by ID?

File: typescript-practice/src/ts/views/product-view.ts [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+  }
+
+  // Validate form modal
+  validator() {
+    this.validateInput();
+    this.displayImage();
+  }
+
+  // Display product
+  renderProduct(products: IProduct[]) {
```
Comment (MinhNguyen41092) [inline typescript-practice/src/ts/views/product-view.ts:208]: This method is too long. Please separate it to smaller ones.

File: typescript-practice/src/ts/views/product-view.ts (hunk 5 above)
Comment (Canh2810) [reply typescript-practice/src/ts/views/product-view.ts:109]: I want to get error message element under each input in form-group so I use class

Comment (MinhNguyen41092) [reply typescript-practice/src/ts/views/product-view.ts:109]: Please assign an meaningful ID for each error message element.

File: typescript-practice/src/ts/views/product-view.ts [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+  }
+
+  // Validate form modal
+  validator() {
+    this.validateInput();
+    this.displayImage();
+  }
+
+  // Create nodes
+  createNodes(product: IProduct) {
```
Comment (MinhNguyen41092) [inline typescript-practice/src/ts/views/product-view.ts]: Wrong method name. You create only one item here.

File: typescript-practice/src/ts/views/product-view.ts [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+    const iconDelete = document.createElement('li');
+    iconDelete.classList.add('fas');
+    iconDelete.classList.add('fa-times');
+    iconDelete.classList.add('delete-product');
+    iconDelete.id = id;
+    btnDeleteProduct.append(iconDelete);
+    productAction.append(btnDeleteProduct);
+    productItem.append(productAction);
+
+    const imgItem = document.createElement('div');
```
Comment (MinhNguyen41092) [inline typescript-practice/src/ts/views/product-view.ts]: Use article to wrap image.

File: typescript-practice/src/ts/views/product-view.ts [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+
+  // Validate form modal
+  validator() {
+    this.validateInput();
+    this.displayImage();
+  }
+
+  createElement(element: string, classes?: Array<string>, id?: string) {
+    const e = document.createElement(element);
+    if(id) e.id = id
```
Comment (MinhNguyen41092) [inline typescript-practice/src/ts/views/product-view.ts:155]: Separate if blocks with empty lines.

File: typescript-practice/src/ts/views/product-view.ts [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+    });
+  }
+
+  // Validate form modal
+  validator() {
+    this.validateInput();
+    this.displayImage();
+  }
+
+  createElement(element: string, classes?: Array<string>, id?: string) {
```
Comment (MinhNguyen41092) [inline typescript-practice/src/ts/views/product-view.ts:152]: Use this format to comment your functions:
```
/**
 * Returns x raised to the n-th power.
 *
 * @param {number} x The number to raise.
 * @param {number} n The power, must be a natural number.
 * @return {number} x raised to the n-th power.
 */
function pow(x, n) {
  ...
}
```

File: typescript-practice/src/ts/views/product-view.ts [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+  bindDeleteAllProduct(handler: Function) {
+    const removeAll = this.getElement('#remove-all');
+    removeAll.addEventListener('click', () => {
+      handler();
+    });
+  }
+
+  // Get Id when selected product
+  bindSelectedProduct(handler: Function) {
+    this.productList.addEventListener('click', (event) => {
```
Comment (MinhNguyen41092) [inline typescript-practice/src/ts/views/product-view.ts]: You bind event to item, not the container.

**Pull request merged**